File: .tmp_py4013/batch_08_dump/2955_number_of_same_end_substrings.js
// LeetCode 2955 - Number of Same-End Substrings
// https://leetcode.com/problems/number-of-same-end-substrings/

var sameEndSubstringCount = function(s, queries) {
    const n = s.length;
    const pref = Array.from({length: n + 1}, () => new Array(26).fill(0));
    for (let i = 0; i < n; i++) {
        for (let c = 0; c < 26; c++) pref[i + 1][c] = pref[i][c];
        pref[i + 1][s.charCodeAt(i) - 97]++;
    }
    const ans = new Array(queries.length);
    for (let qi = 0; qi < queries.length; qi++) {
        const l = queries[qi][0], r = queries[qi][1];
        let total = 0;
        for (let c = 0; c < 26; c++) {
            const cnt = pref[r + 1][c] - pref[l][c];
            total += cnt * (cnt + 1) / 2;
        }
        ans[qi] = total;
    }
    return ans;
};

File: .tmp_rs4013/rust_check13/2955_number_of_same_end_substrings.rs
#![allow(dead_code, unused_variables, unused_mut, unused_assignments, unused_imports)]
struct Solution;
// LeetCode 2955 - Number of Same-End Substrings
// https://leetcode.com/problems/number-of-same-end-substrings/

impl Solution {
    pub fn same_end_substring_count(s: String, queries: Vec<Vec<i32>>) -> Vec<i32> {
        let s = s.as_bytes();
        let n = s.len();
        let mut pref = vec![[0i32; 26]; n + 1];
        for i in 0..n {
            pref[i + 1] = pref[i];
            pref[i + 1][(s[i] - b'a') as usize] += 1;
        }
        let mut ans = vec![0; queries.len()];
        for (qi, q) in queries.iter().enumerate() {
            let l = q[0] as usize;
            let r = q[1] as usize;
            let mut total = 0;
            for c in 0..26 {
                let cnt = pref[r + 1][c] - pref[l][c];
                total += cnt * (cnt + 1) / 2;
            }
            ans[qi] = total;
        }
        ans
    }
}
